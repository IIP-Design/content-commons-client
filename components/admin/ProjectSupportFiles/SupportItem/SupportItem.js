/**
 *
 * SupportItem
 *
 */
import React, {
  useState, useEffect, useRef, useContext,
} from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { graphql } from '@apollo/client/react/hoc';
import compose from 'lodash.flowright';
import { Loader, Popup } from 'semantic-ui-react';
import debounce from 'lodash/debounce';
import Focusable from 'components/Focusable/Focusable';
import ApolloError from 'components/errors/ApolloError';
import GeneralError from 'components/errors/GeneralError/GeneralError';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
// import FileRemoveReplaceButtonGroup from 'components/admin/FileRemoveReplaceButtonGroup/FileRemoveReplaceButtonGroup';
import { LANGUAGES_QUERY } from 'components/admin/dropdowns/LanguageDropdown/LanguageDropdown';
import { getCount } from 'lib/utils';
import { isWindowWidthLessThanOrEqualTo } from 'lib/browser';
import { UploadContext } from '../../ProjectEdit/VideoEdit/UploadContext';


import './SupportItem.scss';

/* eslint-disable react/prefer-stateless-function */
const SupportItem = props => {
  const MAX_FILE_NAME_LENGTH = 45;
  const ITEM_LANG_PROPORTION = 0.3;
  const DELAY_INTERVAL = 1000;

  const listEl = useRef( null );
  const languageEl = useRef( null );
  const mounted = useRef( true );

  const uploadInProgress = useContext( UploadContext );

  const [error, setError] = useState( false );
  const [widths, setWidths] = useState( {
    listItem: 0,
    itemLang: 0,
  } );

  const { item } = props;
  const { listItem, itemLang } = widths;

  const updateWidths = () => {
    if ( mounted.current ) {
      setWidths( {
        listItem: ( listEl.current && listEl.current.offsetWidth ) || 0,
        itemLang: ( languageEl.current && languageEl.current.offsetWidth ) || 0,
      } );
    }
  };

  const debounceResize = debounce( updateWidths, DELAY_INTERVAL );

  const checkFileUrlStatus = async () => {
    if ( item && item.signedUrl ) {
      try {
        const options = {
          headers: {
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            Range: 'bytes=0-0',
          },
        };

        // the head req was not working with the signed url so using get and
        // simulating a head req but only returning 1 byte (see Range header above).
        // Rather hacky so should research a better way
        axios.get( item.signedUrl, options )
          .catch( err => {
            console.dir( err );
            setError( err.isAxiosError );
          } );
      } catch ( err ) {
        console.log( '' );
      }
    }
  };

  const hasError = () => {
    if ( item && typeof item.error !== 'undefined' ) {
      setError( item.error );

      return;
    }
    checkFileUrlStatus();
  };

  useEffect( () => {
    window.addEventListener( 'resize', debounceResize );
    updateWidths();

    return () => {
      window.removeEventListener( 'resize', debounceResize );
      mounted.current = false;
    };
  }, [] );

  useEffect( () => {
    hasError();
  }, [item.signedUrl] );


  if ( !item || !getCount( item ) ) return null;
  /* eslint-disable react/destructuring-assignment */
  if ( props.data.loading ) return 'Loading...';
  if ( props.data.error ) return <ApolloError error={ props.data.error } />;
  /* eslint-enable react/destructuring-assignment */

  /**
   * Truncates long strings with ellipsis
   * @param {string} str the string
   * @param {number} start index for first cutoff point
   * @param {number} end index for ending cutoff point
   * @param {*} replaceWith char to replace subset of string, default is ellipsis
   * @return truncated string
   */
  const getShortFileName = ( str, start, end, replaceWith = '...' ) => (
    `${str.substr( 0, start )}${replaceWith}${str.substr( -end )}`
  );

  /**
   * Determines an integer proportional
   * to a reference number
   * @param {number} reference
   * @return {number}
   */
  const getProportionalNumber = ( reference, proportion ) => (
    Math.floor( reference * proportion )
  );

  const isLongName = ( name = '', maxLength = MAX_FILE_NAME_LENGTH ) => {
    if ( name && typeof name === 'string' ) {
      return name.length > Math.abs( maxLength );
    }

    return false;
  };

  const isLongLang = ( itemWidth, reference, proportion ) => (
    reference > 0 && itemWidth >= getProportionalNumber( reference, proportion )
  );


  const normalizeLanguage = lang => {
    if ( typeof ( lang ) === 'string' ) { // will be object if coming from graphql
      return props.data.languages.filter( l => l.id === lang )[0];
    }

    return lang;
  };

  const normalizeItem = ( prop, defaultValue ) => {
    const fileProp = item[prop];

    if ( fileProp ) {
      return fileProp;
    } if ( item.input ) {
      return item.input[prop.replace( 'file', '' )];
    }

    return defaultValue;
  };

  const filename = normalizeItem( 'filename', '' );
  const filesize = normalizeItem( 'filesize', 0 );
  const language = normalizeLanguage( item.language );
  const strCutoff = isWindowWidthLessThanOrEqualTo( 900 ) ? 12 : 16;
  const shortFileName = getShortFileName( filename, strCutoff, strCutoff );
  const isLongFileName = isLongName( filename );
  const isLongLangName = isLongLang( itemLang, listItem, ITEM_LANG_PROPORTION );
  const isUploading = uploadInProgress && ( item.loaded < filesize );

  const popupStyle = {
    overflowWrap: 'break-word',
    wordWrap: 'break-word',
    msWordBreak: 'break-all',
    wordBreak: 'break-word',
  };

  const renderName = ( str = '', isLang = false ) => {
    if ( ( !isLang && isLongFileName ) || ( isLang && isLongLangName ) ) {
      return (
        <Popup
          content={ str }
          size="mini"
          inverted
          on={ [
            'hover',
            'focus',
          ] }
          trigger={ (
            <span>
              <Focusable>
                { isLang ? str : shortFileName }
              </Focusable>
            </span>
          ) }
          style={ popupStyle }
        />
      );
    }

    return str;
  };

  return (
    <li
      key={ `${item.id}-${language.id}` }
      className={ `support-item ${item.loaded < filesize ? 'new' : ''}` }
      ref={ listEl }
    >
      <span className="item-name" style={ error ? { color: '#cd2026' } : {} }>
        { isLongFileName && <VisuallyHidden>{ filename }</VisuallyHidden> }
        <span
          className={ `item-name-wrap${isLongFileName ? ' hasEllipsis' : ''}` }
          aria-hidden={ isLongFileName }
        >
          { error
            ? <GeneralError msg={ renderName( filename ) } />
            : renderName( filename ) }
        </span>

        <Loader active={ isUploading } inline size="mini" style={ { marginLeft: '.5rem' } } />

      </span>

      <span className={ `item-lang${error ? ' error' : ''}` }>
        { error
          ? (
            null
            /* <FileRemoveReplaceButtonGroup
              onRemove={ () => console.log( 'removed' ) }
            /> */
          )
          : (
            <b
              className={ `item-lang-wrap${isLongLangName ? ' hasEllipsis' : ''}` }
              ref={ languageEl }
            >
              { renderName( language.displayName, true ) }
            </b>
          ) }
      </span>
    </li>
  );
};


SupportItem.propTypes = {
  item: PropTypes.object,
  data: PropTypes.object,
};

export default compose(
  graphql( LANGUAGES_QUERY ),
)( SupportItem );
