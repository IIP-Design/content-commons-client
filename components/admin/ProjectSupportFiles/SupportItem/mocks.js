import { LANGUAGES_QUERY } from 'components/admin/dropdowns/LanguageDropdown/LanguageDropdown';

export const props = {
  item: {
    id: 'cjzts5akb1aj90720hvrhep1t',
    createdAt: '2019-08-27T12:01:34.796Z',
    updatedAt: '2019-08-27T12:01:34.796Z',
    filename: 'image-1.png',
    filetype: 'image/png',
    filesize: 42093736,
    url: '2019/08/cjzts3ele1ahx0720t7u9hf9i/image-1.png',
    signedUrl: 'https://thesignedurl.com',
    alt: null,
    language: {
      id: 'cjsq439dz005607560gwe7k3m',
      locale: 'en-us',
      languageCode: 'en',
      displayName: 'English',
      textDirection: 'LTR',
      nativeName: 'English',
      __typename: 'Language',
    },
    __typename: 'ImageFile',
  },
};

export const preUploadProps = {
  item: {
    language: 'cjsq439dz005607560gwe7k3m',
    use: 'cjtkdq8kr0knf07569goo9eqe',
    quality: '',
    videoBurnedInStatus: '',
    input: {
      lastModified: 1517162382000,
      lastModifiedDate: {},
      name: 'image-1.png',
      size: 42093736,
      type: 'image/png',
      webkitRelativePath: '',
    },
    id: '6ec96575-03e7-4575-af94-05003ddf28ba',
    loaded: 0,
    error: false,
  },
};

export const uploadErrorProps = {
  item: {
    ...preUploadProps.item,
    error: true,
  },
};

export const postUploadProps = {
  item: {
    ...preUploadProps.item,
    loaded: props.item.filesize,
  },
};

export const longFileNameProps = {
  item: {
    ...props.item,
    filename: 'image-image-image-image-image-image-image-1.png',
    url: '2019/08/cjzts3ele1ahx0720t7u9hf9i/image-image-image-image-image-image-image-1.png',
  },
};

export const longLangNameProps = {
  item: {
    ...props.item,
    language: {
      ...props.item.language,
      displayName: 'English-English-English-English-English-English',
    },
  },
};

export const emptyItemProps = {
  item: {},
};

export const nullItemProps = {
  item: null,
};

export const mocks = [
  {
    request: {
      query: LANGUAGES_QUERY,
    },
    result: {
      data: {
        languages: [
          {
            id: 'cjsq439dz005607560gwe7k3m',
            displayName: 'English',
            locale: 'en-us',
            __typename: 'Language',
          },
          {
            id: 'cjsq4565v005c0756f0lqbfe4',
            displayName: 'French',
            locale: 'fr-fr',
            __typename: 'Language',
          },
          {
            id: 'cjsq47fum005i0756qlhc0wta',
            displayName: 'Arabic',
            locale: 'ar',
            __typename: 'Language',
          },
          {
            id: 'cjsq4967o005o0756gvpl1f5k',
            displayName: 'Chinese (Simplified)',
            locale: 'zh-cn',
            __typename: 'Language',
          },
          {
            id: 'cjsuxnpv700dl0756wg4adr6k',
            displayName: 'Spanish',
            locale: 'es-es',
            __typename: 'Language',
          },
          {
            id: 'cjt06xh0e02q70756dnuo3g8e',
            displayName: 'Russian',
            locale: 'ru-ru',
            __typename: 'Language',
          },
          {
            id: 'cjt5qc65o04020756oius55mh',
            displayName: 'Portuguese (Brazil)',
            locale: 'pt-br',
            __typename: 'Language',
          },
          {
            id: 'cjt5qmm4504090756sy269uhs',
            displayName: 'Vietnamese',
            locale: 'vi-vn',
            __typename: 'Language',
          },
          {
            id: 'cjt5qnpem040e07567msmql7j',
            displayName: 'Urdu',
            locale: 'ur',
            __typename: 'Language',
          },
          {
            id: 'cjt5qp676040j07560v969u9p',
            displayName: 'Korean',
            locale: 'ko-kr',
            __typename: 'Language',
          },
          {
            id: 'cjt5qq47h040p0756fjlu2dvy',
            displayName: 'Japanese',
            locale: 'ja-jp',
            __typename: 'Language',
          },
          {
            id: 'cjt5qviod040v0756m5yhxs05',
            displayName: 'Bahasa Indonesia',
            locale: 'id-id',
            __typename: 'Language',
          },
          {
            id: 'cjt5r1hpl0411075624l76vvy',
            displayName: 'Persian',
            locale: 'fa',
            __typename: 'Language',
          },
          {
            id: 'cjubbldh20uqa07560u70yc0k',
            displayName: 'Japanese',
            locale: 'ja',
            __typename: 'Language',
          },
          {
            id: 'cjubbldhj0uqh0756pco0zk2r',
            displayName: 'Persian',
            locale: 'fa-ir',
            __typename: 'Language',
          },
          {
            id: 'cjubbldi10uqj0756oxctum4f',
            displayName: 'Vietnamese',
            locale: 'vi',
            __typename: 'Language',
          },
        ],
      },
    },
  },
];

export const errorMocks = [
  {
    ...mocks[0],
    result: {
      errors: [{ message: 'There was an error.' }],
    },
  },
];

export const emptyMocks = [
  {
    ...mocks[0],
    result: {
      data: { languages: [] },
    },
  },
];

export const nullMocks = [
  {
    ...mocks[0],
    result: {
      data: { languages: null },
    },
  },
];
