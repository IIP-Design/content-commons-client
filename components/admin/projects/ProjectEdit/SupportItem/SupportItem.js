/**
 *
 * SupportItem
 *
 */
import React, { Fragment } from 'react';
import { object, string } from 'prop-types';
import {
  Icon, Loader, Popup, Progress
} from 'semantic-ui-react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import debounce from 'lodash/debounce';

import Focusable from 'components/Focusable/Focusable';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
import './SupportItem.scss';

/* eslint-disable react/prefer-stateless-function */
class SupportItem extends React.PureComponent {
  constructor( props ) {
    super( props );

    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * 1MB = 1,048,576 Bytes
     */
    this.MEGABYTE = 1048576;
    this.MIN_INTERVAL = 500;
    this.MAX_INTERVAL = 1500;
    this.MIN_MB_SEC = 1;
    this.MAX_MB_SEC = 5;
    this.STR_INDEX_PROPORTION = 0.04;
    this.ITEM_NAME_PROPORTION = 0.625;
    this.ITEM_LANG_PROPORTION = 0.3;
    this.DELAY_INTERVAL = 1000;
    this.debounceResize = debounce( this.resetWidths, this.DELAY_INTERVAL );
    this._isMounted = false;

    this.state = {
      bytesUploaded: 0,
      nIntervId: null,
      isUploading: false,
      listItemWidth: null,
      itemNameWidth: null,
      itemLangWidth: null
    };
  }

  componentDidMount = () => {
    this._isMounted = true;
    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * min, max interval in milliseconds
     */
    const interval = this.getRandomInt( this.MIN_INTERVAL, this.MAX_INTERVAL );
    const nIntervId = setInterval( this.uploadItem, interval );
    this.setState( { nIntervId } );

    window.addEventListener( 'resize', this.debounceResize );
  }

  componentWillUnmount = () => {
    this._isMounted = false;
    clearInterval( this.state.nIntervId );
    window.removeEventListener( 'resize', this.debounceResize );
  }

  /**
   * @todo simulate upload for dev purposes;
   * replace for production
   */
  getRandomInt = ( min, max ) => {
    const minNum = Math.ceil( min );
    const maxNum = Math.floor( max );
    return (
      Math.floor( Math.random() * ( ( maxNum - minNum ) + 1 ) ) + minNum
    );
  }

  getRemainingUnits = ( totalUnits, unitsUploaded ) => (
    totalUnits - unitsUploaded
  )

  /**
   * Truncates long strings with ellipsis
   * @param {string} str the string
   * @param {number} start index for first cutoff point
   * @param {number} end index for ending cutoff point
   * @return truncated string
   */
  getShortFileName = ( str, index ) => (
    <Fragment>
      { `${str.substr( 0, index )}` }&hellip;{ `${str.substr( -index )}` }
    </Fragment>
  );

  /**
   * Determines an integer proportional
   * to a reference number
   * @param {number} reference
   * @return {number}
   */
  getProportionalNumber = ( reference, proportion ) => (
    Math.floor( reference * proportion )
  )

  /**
   * Declares a React ref & sets its width in state
   * @param {node} React node
   * @param {string} name of React ref
   */
  setRefWidth = ( node, ref ) => {
    if ( node ) {
      this.setState( prevState => {
        if ( !prevState[`${ref}Width`] ) {
          return ( {
            [`${ref}Width`]: Math.ceil( node.offsetWidth )
          } );
        }
      } );
    }
  }

  resetWidths = () => {
    if ( this._isMounted ) {
      this.setState( {
        listItemWidth: null,
        itemNameWidth: null,
        itemLangWidth: null
      } );
    }
  }

  isLongName = ( itemWidth, reference, proportion ) => (
    itemWidth >= this.getProportionalNumber( reference, proportion )
  );

  incrementUpload = ( unit, min, max ) => (
    this[unit] * this.getRandomInt( min, max )
  )

  endUpload = intervId => {
    clearInterval( intervId );
  };

  uploadItem = () => {
    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * min, max increment in megabytes
     */
    this.setState( nextState => {
      const { filesize } = this.props.data.supportItem;
      const { bytesUploaded } = nextState;
      let increment = this.incrementUpload( 'MEGABYTE', this.MIN_MB_SEC, this.MAX_MB_SEC );
      const remainingBytes = this.getRemainingUnits( filesize, bytesUploaded );

      if ( remainingBytes < increment ) {
        increment = remainingBytes;
      }

      // continue uploading
      if ( bytesUploaded < filesize ) {
        return {
          isUploading: true,
          bytesUploaded: this.state.bytesUploaded + increment
        };
      }

      // stop uploading
      this.endUpload( nextState.nIntervId );
      return { isUploading: false };
    } );
  }

  render() {
    const {
      fileType,
      data: { error, loading, supportItem }
    } = this.props;

    if ( loading ) {
      return (
        <li>
          <Loader active inline size="mini" />
          <span style={ { marginLeft: '0.5em', fontSize: '0.75em' } }>
            Preparing file for upload...
          </span>
        </li>
      );
    }

    if ( error ) {
      return (
        <li className="support-item error">
          <p>
            <Icon
              color="red"
              name="exclamation triangle"
              size="small"
            />
            <span>
              { `${error ? 'Loading ' : 'Uploading '}error` }
            </span>
          </p>
        </li>
      );
    }

    if ( !supportItem || !Object.keys( supportItem ).length ) return null;

    const { filename, filesize, language } = supportItem;

    const {
      bytesUploaded,
      isUploading,
      listItemWidth,
      itemNameWidth,
      itemLangWidth
    } = this.state;

    const uploadingClass = isUploading ? ' isUploading' : '';

    const charIndex = this.getProportionalNumber( listItemWidth, this.STR_INDEX_PROPORTION );

    const shortFileName = this.getShortFileName( filename, charIndex );

    const isLongFileName = this.isLongName( itemNameWidth, listItemWidth, this.ITEM_NAME_PROPORTION );

    const isLongLangName = this.isLongName( itemLangWidth, listItemWidth, this.ITEM_LANG_PROPORTION );

    const popupStyle = {
      overflowWrap: 'break-word',
      wordWrap: 'break-word',
      msWordBreak: 'break-all',
      wordBreak: 'break-word'
    };

    if ( isUploading ) {
      return (
        <li key={ `${fileType}-${language.displayName}` } className={ `support-item${uploadingClass}` }>
          <Progress
            value={ bytesUploaded }
            total={ filesize }
            color="blue"
            size="small"
            active
            progress
            precision={ 0 }
          />
          <span>Uploading item</span>
        </li>
      );
    }

    return (
      <li
        key={ `${fileType}-${language.displayName}` }
        className="support-item"
        ref={ node => this.setRefWidth( node, 'listItem' ) }
      >
        <span className="item-name">
          { isLongFileName && <VisuallyHidden>{ filename }</VisuallyHidden> }
          <span
            className={
              `item-name-wrap${isLongFileName ? ' hasEllipsis' : ''}`
            }
            aria-hidden={ isLongFileName }
            ref={ node => this.setRefWidth( node, 'itemName' ) }
          >
            { isLongFileName
              ? (
                <Popup
                  content={ filename }
                  size="mini"
                  inverted
                  on={ [
                    'hover',
                    'click',
                    'focus'
                  ] }
                  trigger={ (
                    <span>
                      <Focusable>{ shortFileName }</Focusable>
                    </span>
                  ) }
                  style={ popupStyle }
                />
              ) : filename }
          </span>
        </span>

        <span className="item-lang">
          <b
            className={
              `item-lang-wrap${isLongLangName ? ' hasEllipsis' : ''}`
            }
            ref={ node => this.setRefWidth( node, 'itemLang' ) }
          >
            { isLongLangName
              ? (
                <Popup
                  trigger={ (
                    <span>
                      <Focusable>{ language.displayName }</Focusable>
                    </span>
                  ) }
                  content={ language.displayName }
                  on={ [
                    'hover',
                    'click',
                    'focus'
                  ] }
                  inverted
                  size="mini"
                  style={ popupStyle }
                />
              ) : language.displayName }
          </b>
        </span>
      </li>
    );
  }
}

SupportItem.propTypes = {
  data: object.isRequired,
  fileType: string.isRequired
};

const SUPPORT_ITEM_QUERY = gql`
  query SupportItem($id: ID!) {
    supportItem: supportFile(id: $id) {
      id
      filename
      filesize
      language {
        displayName
      }
    }
  }
`;

export default graphql( SUPPORT_ITEM_QUERY, {
  options: props => ( {
    variables: {
      id: props.itemId
    },
  } )
} )( SupportItem );

export { SUPPORT_ITEM_QUERY };
