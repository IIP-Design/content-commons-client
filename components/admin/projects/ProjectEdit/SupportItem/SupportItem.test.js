import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { Icon, Loader } from 'semantic-ui-react';
import SupportItem, { SUPPORT_ITEM_QUERY } from './SupportItem';

const props = {
  projectId: '123',
  fileType: 'srt',
  itemId: 'qwqw'
};

const mocks = [
  {
    request: {
      query: SUPPORT_ITEM_QUERY,
      variables: { id: props.itemId }
    },
    result: {
      data: {
        supportItem: {
          id: props.itemId,
          filename: 'madeinamerica_english.srt',
          filesize: '24576',
          language: {
            displayName: 'English'
          }
        }
      }
    }
  }
];

const resizeWindow = ( x, y ) => {
  window.innerWidth = x;
  window.innerHeight = y;
  window.dispatchEvent( new Event( 'resize' ) );
};

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <SupportItem { ...props } />
  </MockedProvider>
);

describe( '<SupportItem />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const supportItem = wrapper.find( 'SupportItem' );
    const loader = <Loader active inline size="mini" />;

    expect( supportItem.exists() ).toEqual( true );
    expect( supportItem.contains( loader ) ).toEqual( true );
    expect( supportItem.find( 'span' ).text() )
      .toEqual( 'Preparing file for upload...' );
    expect( toJSON( supportItem ) ).toMatchSnapshot();
  } );

  it( 'renders error message & icon if error is thrown', async () => {
    const errorMocks = [
      {
        request: {
          query: SUPPORT_ITEM_QUERY,
          variables: { id: props.itemId }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <SupportItem { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const li = supportItem.find( 'li.support-item.error' );
    const icon = <Icon color="red" name="exclamation triangle" size="small" as="i" />;
    const span = <span>Loading error</span>;

    expect( supportItem.exists() ).toEqual( true );
    expect( li.exists() ).toEqual( true );
    expect( supportItem.contains( icon ) ).toEqual( true );
    expect( supportItem.contains( span ) ).toEqual( true );
  } );

  it( '`componentDidMount` sets `_isMounted`', async () => {
    const wrapper = mount( Component );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    inst.componentDidMount();
    expect( inst._isMounted ).toEqual( true );
  } );

  it( '`componentWillUnmount` sets `_isMounted` to false', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    inst.componentWillUnmount();
    expect( inst._isMounted ).toEqual( false );
  } );

  it( 'renders `Progress` component if `isUploading`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = () => wrapper.find( 'SupportItem' );
    const progress = () => supportItem().find( 'Progress' );
    const isUploading = () => supportItem().state( 'isUploading' );
    const filesize = () => supportItem().prop( 'data' ).supportItem.filesize;

    expect( isUploading() ).toEqual( false );
    expect( progress().exists() ).toEqual( isUploading() );

    supportItem().setState( { isUploading: true } );

    expect( isUploading() ).toEqual( true );
    expect( progress().exists() ).toEqual( isUploading() );
    expect( progress().prop( 'total' ) ).toEqual( filesize() );
    expect( toJSON( supportItem() ) ).toMatchSnapshot();
  } );

  it( 'renders null if `supportItem` is null', async () => {
    const nullMocks = [
      {
        request: {
          query: SUPPORT_ITEM_QUERY,
          variables: { id: props.itemId }
        },
        result: { data: { supportItem: null } }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename={ false }>
        <SupportItem { ...props } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );

    expect( supportItem.html() ).toEqual( null );
  } );

  it( 'calls `debounceResize` when window is resized', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();
    const spy = jest.spyOn( inst, 'debounceResize' );

    inst.componentDidMount();

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    resizeWindow( 900, 600 );
    expect( spy ).toHaveBeenCalled();
  } );

  it( '`componentWillUnmount` removes the `resize` event listener', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    const map = {};
    window.removeEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    inst.componentWillUnmount();
    expect( window.removeEventListener )
      .toHaveBeenCalledWith( 'resize', inst.debounceResize );
  } );

  it( '`getShortFileName` returns an ellipsis shortened file name', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    const longName = 'madeinamerica_chinese_ljhlkjhl_kjhlkjh_aslkfja;lskjfweoij.srt';
    const index = 8;
    const shortName = inst
      .getShortFileName( longName, index )
      .props
      .children;

    expect( shortName ).toHaveLength( 3 );
    expect( shortName[0] ).toHaveLength( 8 );
    expect( shortName[1] ).toHaveLength( 1 );
    expect( shortName[2] ).toHaveLength( 8 );
    expect( shortName.join( '' ) ).toEqual( 'madeinamâ€¦eoij.srt' );
  } );

  it( '`getProportionalNumber` returns a number that is proportional to a reference', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    const reference = 500;
    const proportion = 0.625;
    const num = inst
      .getProportionalNumber( reference, proportion );
    expect( num ).toEqual( 312 );
  } );

  it( '`isLongName` returns a boolean whether given number is greater or equal to another', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    let itemWidth = 290;
    const reference = 500;
    const proportion = 0.625;

    const isLongName = () => (
      inst.isLongName( itemWidth, reference, proportion )
    );

    expect( isLongName() ).toEqual( false );
    itemWidth = 340;
    expect( isLongName() ).toEqual( true );
  } );

  it( '`setRefWidth` and `resetWidths` sets ref widths in state', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    const li = supportItem.find( '.support-item' );
    const span = supportItem.find( '.item-name-wrap' );
    const b = supportItem.find( '.item-lang-wrap' );

    const nodes = [li, span, b];
    const names = ['listItem', 'itemName', 'itemLang'];

    li.offsetWidth = 200;
    span.offsetWidth = 170;
    b.offsetWidth = 50;

    nodes.forEach( ( n, i ) => {
      expect( supportItem.state( `${names[i]}Width` ) )
        .toEqual( 0 );
    } );

    nodes.forEach( ( n, i ) => {
      inst.setRefWidth( n, names[i] );
      expect( supportItem.state( `${names[i]}Width` ) )
        .toEqual( n.offsetWidth );
    } );

    inst.resetWidths();

    nodes.forEach( ( n, i ) => {
      expect( supportItem.state( `${names[i]}Width` ) )
        .toEqual( 0 );
    } );
  } );

  it( '`resetWidths` does not reset ref widths in state if component is unmounted', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const supportItem = wrapper.find( 'SupportItem' );
    const inst = supportItem.instance();

    const li = supportItem.find( '.support-item' );
    const span = supportItem.find( '.item-name-wrap' );
    const b = supportItem.find( '.item-lang-wrap' );

    const nodes = [li, span, b];
    const names = ['listItem', 'itemName', 'itemLang'];

    li.offsetWidth = 200;
    span.offsetWidth = 170;
    b.offsetWidth = 50;

    inst.componentDidMount();
    const isMounted = () => inst._isMounted;

    expect( isMounted() ).toEqual( true );

    nodes.forEach( ( n, i ) => {
      expect( supportItem.state( `${names[i]}Width` ) )
        .toEqual( 0 );
    } );

    nodes.forEach( ( n, i ) => {
      inst.setRefWidth( n, names[i] );
      expect( supportItem.state( `${names[i]}Width` ) )
        .toEqual( n.offsetWidth );
    } );

    inst.componentWillUnmount();
    expect( isMounted() ).toEqual( false );

    inst.resetWidths();

    nodes.forEach( ( n, i ) => {
      expect( supportItem.state( `${names[i]}Width` ) )
        .toEqual( n.offsetWidth );
    } );
  } );
} );
