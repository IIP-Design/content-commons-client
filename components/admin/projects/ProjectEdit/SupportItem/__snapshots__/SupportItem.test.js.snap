// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SupportItem /> renders \`Progress\` component if \`isUploading\` 1`] = `
<SupportItem
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "supportItem": Object {
        "filename": "madeinamerica_english.srt",
        "filesize": "24576",
        "id": "qwqw",
        "language": Object {
          "displayName": "English",
        },
      },
      "updateQuery": [Function],
      "variables": Object {
        "id": "qwqw",
      },
    }
  }
  fileType="srt"
  itemId="qwqw"
  projectId="123"
>
  <li
    className="support-item isUploading"
    key="srt-English"
  >
    <Progress
      active={true}
      color="blue"
      precision={0}
      progress={true}
      size="small"
      total="24576"
      value={0}
    >
      <div
        className="ui blue small active progress"
        data-percent={0}
      >
        <div
          className="bar"
          style={
            Object {
              "width": "0%",
            }
          }
        >
          <div
            className="progress"
          >
            0%
          </div>
        </div>
      </div>
    </Progress>
    <span>
      Uploading item
    </span>
  </li>
</SupportItem>
`;

exports[`<SupportItem /> renders initial loading state without crashing 1`] = `
<SupportItem
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "qwqw",
      },
    }
  }
  fileType="srt"
  itemId="qwqw"
  projectId="123"
>
  <li>
    <Loader
      active={true}
      inline={true}
      size="mini"
    >
      <div
        className="ui mini active inline loader"
      />
    </Loader>
    <span
      style={
        Object {
          "fontSize": "0.75em",
          "marginLeft": "0.5em",
        }
      }
    >
      Preparing file for upload...
    </span>
  </li>
</SupportItem>
`;
